src/core/common/retain_ptr.hpp

```diff
@@ -0,0 +1,286 @@
+ *
+ * @tparam Type  The pointer type.
+ *
+ */
+template <class Type> class RetainPtr
+{
+public:
+    /**
+     * This is the default constructor for `RetainPtr` initializing it as null.
+     *
+     */
+    RetainPtr(void)
+        : mPointer(nullptr)
+    {
+    }
+
+    /**
+     * This constructor initializes the `RetainPtr` with a given pointer.
+     *
+     * Upon construction the `RetainPtr` will increment the retain count on @p aPointer (if not null).
+     *
+     * @param[in] aPointer  A pointer to object to initialize with.
+     *
+     */
+    explicit RetainPtr(Type *aPointer)
+        : mPointer(aPointer)
+    {
+        IncrementRetainCount();
+    }
+
+    /**
+     * This constructor initializes the `RetainPtr` from another `RetainPtr`.
+     *
+     * @param[in] aOther   Another `RetainPtr`.
+     *
+     */
+    RetainPtr(const RetainPtr &aOther)
+        : mPointer(aOther.mPointer)
+    {
+        IncrementRetainCount();
+    }
+
+    /**
+     * This is the destructor for `RetainPtr`.
+     *
+     * Upon destruction, the `RetainPtr` will decrement the retain count on the managed object (if not null) and
+     * free the object if its retain count reaches zero.
+     *
+     */
+    ~RetainPtr(void) { DecrementRetainCount(); }
+
+    /**
+     * This method indicates whether the `RetainPtr` is null or not.
+     *
+     * @retval TRUE   The `RetainPtr` is null.
+     * @retval FALSE  The `RetainPtr` is not null.
+     *
+     */
+    bool IsNull(void) const { return (mPointer == nullptr); }
+
+    /**
+     * This method gets the raw pointer to the object managed by `RetainPtr`.
+     *
+     * @returns The raw pointer to the object managed by `RetainPtr` or `nullptr` if none.
+     *
+     */
+    Type *Get(void) { return mPointer; }
+
+    /**
+     * This method gets the raw pointer to the object managed by `RetainPtr`.
+     *
+     * @returns The raw pointer to the object managed by `RetainPtr` or `nullptr` if none.
+     *
+     */
+    const Type *Get(void) const { return mPointer; }
+
+    /**
+     * This method replaces the managed object by `RetainPtr` with a new one.
+     *
+     * The method correctly handles a self `Reset()` (i.e., @p aPointer being the same pointer as the one currently
+     * managed by `RetainPtr`).
+     *
+     * @param[in] aPointer   A pointer to a new object to replace with.
+     *
+     */
+    void Reset(Type *aPointer = nullptr)
+    {
+        if (aPointer != mPointer)
+        {
+            DecrementRetainCount();
+            mPointer = aPointer;
+            IncrementRetainCount();
+        }
+    }
+
+    /**
+     * This method releases the ownership of the current pointer in `RetainPtr` (if any) without changing its retain
+     * count.
+     *
+     * After this call, the `RetainPtr` will be null.
+     *
+     * @returns The pointer to the object managed by `RetainPtr` or `nullptr` if `RetainPtr` was null.
+     *
+     */
+    Type *Release(void)
```

Comment:
Seems the basic idea is that each live `RetainPtr` instance will increase the `Retain Count` of the underlying `RetainCountable` object by exactly once. 
However, the `Release` method violates this rule by detaching the pointer without reducing the `Retain Count`. 

I think we should remove `Release` method to avoid careless mistakes unless there is a clear use case for this method. 

Thoughts? @abtink @wgtdkp

Reply:
I agree `Release()` should be used with caution for special situations.

The main use-case of `Release()` is when interfacing with C APIs (or other modules) which do not know/use reference counted (or smart) pointers. We want to detach the pointer from being managed by `RetainPtr` and just get back the ownership of the raw pointer.

Reply:
> We want to detach the pointer from being managed by RetainPtr and just get back the ownership of the raw pointer.

We can always use another `RetainPtr` to retain the ownership safely?

Reply:
It can be useful if we want to pass an allocated object (pass the object and its ownership) to the user in an OT public API (which is C based and does not have notion of smart pointers or `RetainPtr` and works with raw pointer).  

I know some libraries choose to not provide this for the fear of being incorrectly used (which make sense for more general purpose libraries like STL that can be used by developers of all levels of experience).  For example,  `std::shared_ptr` does not have it, but [`boost::intrusive_ptr`](https://www.boost.org/doc/libs/1_60_0/libs/smart_ptr/intrusive_ptr.html) provides `detach()` and also the [`retain_ptr`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0468r1.html) proposal for STL  provides this.  I think within OT we can make sure that it is not incorrectly used.